### PR TITLE
IOC hardening

### PR DESCRIPTION
**Pull request by anjohnson**
A collection of hardening fixes, mostly fixed or reported by [Jacqueline Garrahan](https://github.com/jacquelinegarrahan) in #912 but with amendments or alternative fixes and added tests by me (assisted by Claude Opus 4.8 and friends).

### CODE REVIEW

File: modules/database/src/ioc/as/asDbLib.c [hunk 1]
```diff
@@ -157,6 +157,10 @@ int asInit(void)
 int asShutdown(void) {
     volatile ASBASE *pbase = pasbase;
     pasbase = NULL;
+    /* restore the invariant that asActive implies pasbase != NULL; leaving
+       asActive TRUE here would let later AS calls dereference the freed/NULL
+       base or consult stale cached access rights */
+    asActive = FALSE;
```
Comment (mdavidsaver) [inline modules/database/src/ioc/as/asDbLib.c:163]: A reasonable commit message, but imo. inappropriate verbosity in code.  The comment added near the definitions of these two globals is the correct place to document this kind of relationship, not at all/some usages.

File: modules/database/src/ioc/dbStatic/dbStaticLib.c [hunk 2]
```diff
@@ -2591,14 +2591,16 @@ long dbPutString(DBENTRY *pdbentry,const char *pstring)
         ((char *)pfield)[pflddes->size-1] = 0;
 
         if((pflddes->special == SPC_CALC) && !stringHasMacro) {
-            char  rpcl[RPCL_LEN];
+            /* size the postfix output from the field's actual capacity */
+            char *rpcl = dbmfMalloc(INFIX_TO_POSTFIX_SIZE(pflddes->size));
```
Comment (mdavidsaver) [inline modules/database/src/ioc/dbStatic/dbStaticLib.c:2595]: This function can be called after `iocInit()`.  Is `dbmfMalloc()` the appropriate allocator?

The added comment is redundant.

File: modules/database/src/ioc/dbStatic/dbStaticLib.c [hunk 3]
```diff
@@ -2746,10 +2748,12 @@ char * dbVerify(DBENTRY *pdbentry, const char *pstring)
         }
 
         if (pflddes->special == SPC_CALC) {
-            char  rpcl[RPCL_LEN];
+            /* size the postfix output from the field's actual capacity */
+            char *rpcl = dbmfMalloc(INFIX_TO_POSTFIX_SIZE(pflddes->size));
```
Comment (mdavidsaver) [inline modules/database/src/ioc/dbStatic/dbStaticLib.c:2752]: ditto

File: modules/database/test/std/rec/printfTest.c [hunk 4]
```diff
@@ -522,6 +524,70 @@ static void test_all_inputs(void){
...
+    /* A '%' directive longer than doPrintf()'s internal format[20] buffer must
+     * be rejected with F_BADFMT. The accepted prefix is echoed and the
+     * remaining characters copied as literal text, so the trailing conversion
+     * is NOT applied. A too-long "%...d" therefore yields a '#' string, not
+     * the formatted value "0". */
+    char format_string[MAX_STRING_SIZE];
+    char result_string[MAX_STRING_SIZE];
+
+    format_string[0] = '%';
+    memset(format_string + 1, '#', 25);     /* well past the 20-byte limit */
```
Comment (mdavidsaver) [inline modules/database/test/std/rec/printfTest.c:537]: Would it be useful to test with other filler characters?  eg. `'0'`?

File: modules/libcom/src/as/asLibRoutines.c [hunk 5]
```diff
@@ -38,6 +38,13 @@ static epicsMutexId asLock;
 #define LOCK epicsMutexMustLock(asLock)
 #define UNLOCK epicsMutexUnlock(asLock)
 
+/* pasbase points to the currently active access-security configuration, or
+ * NULL when none is loaded.  asActive is set TRUE once a configuration has
+ * been successfully loaded and must only be TRUE while pasbase is non-NULL;
+ * when FALSE the asCheck* macros default to granting access.  Code that tears
+ * down the configuration (eg. asShutdown) must clear asActive before/when it
+ * clears pasbase so the two stay consistent.
```
Comment (mdavidsaver) [inline modules/libcom/src/as/asLibRoutines.c:46]: imo. too wordy.  I would say rather:

```c
/* Allowed states:
 *   pasbase==NULL && !asActive
 *   pasbase!=NULL && asActive
 * Guarded by asLock
 */
```

**Comment by hjunkes** [comment]
Ran this PR through end-to-end testing today against our RTEMS 7 BeagleBone Black IOC (`beaglePyroIOC`), cross-compiled and hardware-verified on a real board (not just host/CI). AI-assisted (Claude Code), fully reviewed and run by me.

**What was tested:**
- Cherry-picked all 7 commits onto our BBB fork's `epics-base` `7.0` branch.
- Host build (darwin-aarch64): full `libcom` + `database` module test suites (10k+ tests) pass, including this PR's own new `printfTest` cases.
- Added 3 extra regression tests on top: a `'0'`-filled `printfRecord` format-overflow variant (per @mdavidsaver's review comment asking about other fill characters), and two `postfix()` cases (a 220-character chain of unary minus, and 150 levels of nested parens) that would have overrun the old fixed `ELEMENT stack[80]`.
- Cross-compiled for `RTEMS-beagleboneblack` and booted on real hardware (no MMU, no stack canaries — the class of environment this hardening actually matters for).
- Interactive iocsh regression records (`hardeningTest.db`) exercising the two fixes reachable from this IOC:
  - Boot-time: three `calc` records with `CALC` set directly via `dbLoadRecords` (the `dbStaticLib.c` `dbPutString`/`dbVerify` path) — a 150x unary-minus chain, 70 nested parens, and a 159-char dense expression filling the field's full capacity.
  - Live/runtime: a `calc` record re-poked via `.CALC$` and a `printf` record re-poked via `.FMT`, both exercising `postfix()` and `doPrintf()` repeatedly without rebooting.

**Results:** all matched predicted output exactly, no crashes or hangs —
- Boot-time CALC records: `VAL` = 1, 1, 80 as expected.
- Live `postfix()` retest: `VAL` = 1 as expected.
- Live `printfRecord` retest: `.VAL` = `"Test: %####################d"` (18 `#` accepted into `format[20]`, 19th triggers the bound check and is dropped, remainder copied as literal text) — exact match.

Not exercised (this IOC loads no `.acf`): the `asShutdown`/`asActive` fix and the `asLib_lex.l` bound fix. Logically reviewed, covered by the PR's own scope, just nothing here triggers them.

Happy to share the `hardeningTest.db` / test records if useful to anyone else validating this on RTEMS or other constrained targets. Links:
- Interactive hardware test records (`hardeningTest.db`): [commit](https://github.com/hjunkes/beaglePyroIOC/commit/ca168ec5e0f9bc5e1042802c950d7e8abb222cb3) / [file](https://github.com/hjunkes/beaglePyroIOC/blob/ca168ec5e0f9bc5e1042802c950d7e8abb222cb3/beaglePyroIOCApp/Db/hardeningTest.db)
- Added `printfTest`/`epicsCalcTest` regression tests: [commit](https://github.com/hjunkes/epics-base/commit/2e5342f0e4b6b42cf762a9b3d9e005710520fae4) ([diff](https://github.com/hjunkes/epics-base/commit/2e5342f0e4b6b42cf762a9b3d9e005710520fae4.diff))
- Whole branch (this PR cherry-picked + the above, on top of our BBB RTEMS port): [hjunkes/epics-base@7.0](https://github.com/hjunkes/epics-base/tree/7.0)

One question: was this kind of on-target hardware regression testing something that was requested in yesterday's video conference? Want to make sure this covers what was actually asked for rather than just what seemed useful to me.
